lib/vega_lite/data.ex

```diff
@@ -116,31 +137,33 @@ defmodule VegaLite.Data do
   defp annotated_heatmap(vl, data, fields) do
     text_fields = [text: fields[:text], x: fields[:x], y: fields[:y]]
     fields = List.keydelete(fields, :text, 0)
-    Vl.layers(vl, [chart(data, :rect, fields), chart(data, :text, text_fields)])
+    used_fields = fields |> Keyword.values() |> used_fields()
+
+    vl
+    |> Vl.data_from_values(data, only: used_fields)
+    |> Vl.layers([layer(data, :rect, fields), layer(data, :text, text_fields)])
   end
 
-  defp heatmap_defaults({field, {col, opts}}) when field in [:x, :y] do
-    opts = Keyword.put_new(opts, :type, :nominal)
-    {field, {col, opts}}
+  defp heatmap_defaults({field, [{:field, _col} | _] = opts}) when field in [:x, :y] do
+    {field, Keyword.put_new(opts, :type, :nominal)}
   end
 
   defp heatmap_defaults({field, col}) when field in [:x, :y] do
-    {field, {col, type: :nominal}}
+    {field, [field: col, type: :nominal]}
   end
 
-  defp heatmap_defaults({field, {col, opts}}) when field in [:color, :text] do
-    opts = Keyword.put_new(opts, :type, :quantitative)
-    {field, {col, opts}}
+  defp heatmap_defaults({field, [{:field, _col} | _] = opts}) when field in [:color, :text] do
+    {field, Keyword.put_new(opts, :type, :quantitative)}
   end
 
   defp heatmap_defaults({field, col}) when field in [:color, :text] do
-    {field, {col, type: :quantitative}}
+    {field, [field: col, type: :quantitative]}
   end
 
-  defp enconde_mark(vl, {mark, opts}), do: Vl.mark(vl, mark, opts)
+  defp enconde_mark(vl, [{:type, mark} | opts]), do: Vl.mark(vl, mark, opts)
```

Comment:
We cannot assume type is the first element, so you need something like:

```suggestion
  defp enconde_mark(vl, opts) when is_list(opts) do
    {mark, opts} = Keyword.pop!(opts, :type)
    Vl.mark(vl, mark, opts)
  end
```